solver/src/main/java/org/chocosolver/solver/constraints/extension/hybrid/PropHybridTable.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * This file is part of choco-solver, http://choco-solver.org/
+ *
+ * Copyright (c) 2023, IMT Atlantique. All rights reserved.
+ *
+ * Licensed under the BSD 4-clause license.
+ *
+ * See LICENSE file in the project root for full license information.
+ */
+package org.chocosolver.solver.constraints.extension.hybrid;
+
+import org.chocosolver.memory.IEnvironment;
+import org.chocosolver.memory.IStateInt;
+import org.chocosolver.solver.ICause;
+import org.chocosolver.solver.constraints.Propagator;
+import org.chocosolver.solver.constraints.PropagatorPriority;
+import org.chocosolver.solver.exception.ContradictionException;
+import org.chocosolver.solver.variables.IntVar;
+import org.chocosolver.util.ESat;
+import org.chocosolver.util.objects.setDataStructures.ISet;
+import org.chocosolver.util.objects.setDataStructures.SetFactory;
+import org.chocosolver.util.objects.setDataStructures.SetType;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+
+/**
+ * This class implements an adaptation of the filtering algorithm from:
+ * "Extending Compact-Table to Basic Smart Tables",
+ * H. Verhaeghe and C. Lecoutre and Y. Deville and P. Schauss, CP-17.
+ * to STR2. It deals with smart/hybrid tuples.
+ * <br/>
+ *
+ * @author Charles Prud'homme
+ * @since 13/02/2023
+ */
+public class PropHybridTable extends Propagator<IntVar> {
+
+    private final HReExpression[][] table;
+    private final StrHVar[] str2vars;
+    private final ISet activeTuples;
+    private final ArrayList<StrHVar> ssup;
+    private final ArrayList<StrHVar> sval;
+    private boolean firstProp = true;
+
+    public PropHybridTable(IntVar[] vars, HybridTuples tuples) {
+        super(vars, PropagatorPriority.QUADRATIC, false);
+        this.table = tuples.toArray();
+        int size = 0;
+        if (table.length > 0) {
+            size = table[0].length;
+        }
+        str2vars = new StrHVar[size];
+        int max = 0;
+        for (int i = 0; i < size; i++) {
+            str2vars[i] = new StrHVar(model.getEnvironment(), vars[i], i);
+            max = Math.max(max, vars[i].getUB());
```

Comment:
What is the purpose of max?

Reply:
That's an undesirable copy/paste. To be removed